---
title: "GRNIToolS"
author: "Yongqiang Zhou"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{GRNIToolS}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Introduction

In this work, we present a comprehensive open source R package of Gene Regulatory Network Inference Tools Selector (GRNIToolS) with benchmarking real biological data from high-throughput Omics experiments to compare and assess the performance of various algorithms on gene regulatory network inference. This package contains of several inference algorithms that cover models of Mutual Information, Correlation, Machine learning, Meta algorithms, including microarray and single-cell RNA-seq data of mouse embryonic stem cells. This package provides modules that can be input into gold standards as well as algorithms, which is helpful for the rigorous and scalable evaluation of the GRN construction methods under the specific data.


# Example for main GRNIToolS
## Preparing data
Expression matrix B and gene list need to be prepared. A random set of data is generated to demonstrate how to use this package.
```{r step1 data prepare}
library(GRNIToolS)
ntf <- 30
ngene <- 100
nsample <- 10
set.seed(123)
A <- matrix(rnorm(ntf * nsample,0,1),ntf,nsample)
TF <- paste0('G',1:ntf)
B <- matrix(rnorm(ngene * nsample,0,1),ngene,nsample)
X <- matrix(0,ngene,ntf)
s <- c(1:10)
gene <- paste0('G',1:ngene)
HGSmat <- matrix(0,ngene,ntf)
HGSmat[sample(ngene * ntf,0.2 * ngene * ntf)] <- 1
colnames(HGSmat) <- TF
rownames(HGSmat) <- gene
```

## Network inference
Select the network inference method.If the regression method is selected, additional matrix A and matrix X need to be provided, and other methods do not need this datas.
```{r step2 run_grn, echo=TRUE}
res <- demo_reg(A, B, s, X, 'ADMM',max.steps = 200,TF, gene, cl.cores = 2,file=NULL,verbose=TRUE)
grn <- res$grn 
grn[1:5,1:5]
res <- demo_other(B,'Pearson',TF=TF,gene=gene,file=NULL,verbose=TRUE)
grn <- res$grn 
grn[1:5,1:5]
```

## Evalution
Calculate evaluation parameters FP, FN, Precision, Recall, F1-score, AUROC and AUPR.
```{r step3 evalution}
dat <- calpl(HGSmat = HGSmat,grn = grn,sym = F)
confu <- confusion(dat = dat,threshold = 0.5)
print(confu)
auc <- calauc(dat = dat)
print(auc)
```
## ROC and PR curve
```{r step4 ROC and PR curve, echo=TRUE, fig.height=5, fig.width=6.5, warning=FALSE}
p1 <- draw_plot(dat,method = 'ROC',algname = 'Pearson')
p2 <- draw_plot(dat,method = 'PR',algname = 'Pearson')
print(p1)
print(p2)
```

## Network diagram
Generate a visual network diagram and generate a data format that can be used for Cytoscape to establish the network.
```{r step5 output or draw network}
net <- network(grn,TF = TF,gene = gene)
#print(net)
grn <- out_grn(grn,TF = TF,gene = gene)
head(grn)
```


# The pipeline for GRNIToolS to evalute this methods
```{r pipeline, echo=TRUE, fig.height=7, fig.width=9}
algorithm <- c('Pearson','parmigene_ARACNE.a','minet_MRNET','PCIT','ADMM')
conlist <- grn_main(A=A,B=B,X=X,seq = s,gene=gene,TF=TF,algorithm = algorithm,
                    HGSmat = HGSmat,file=TRUE, verbose=FALSE,bootstrap_num = 5,sym = FALSE,threshold = 0.5)
print(conlist[1:10,])
dat <- data.frame(alg = conlist$Algorithm,auc = conlist$AUROC)
p <- draw_box(dat,'AUROC')
print(p)
```

